Content.Server/Climbing/ClimbSystem.cs

```diff
@@ -1,121 +1,445 @@
+    /// </summary>
+    /// <returns>Returns whether adding the new fixtures was successful</returns>
+    private bool ReplaceFixtures(ClimbingComponent climbingComp, PhysicsComponent physicsComp, FixturesComponent fixturesComp)
+    {
+        // Swap fixtures
+        var toAdd = new List<Fixture>();
+        foreach (var (name, fixture) in fixturesComp.Fixtures)
         {
-            if (!Resolve(uid, ref component, false))
-                return;
-            component.IsClimbing = true;
-            UnsetTransitionBoolAfterBufferTime(uid, component);
+            if (climbingComp.Fixtures.Contains(name)
+                || climbingComp.Fixtures.Contains($"{ClimbingFixtureName}-{name}")
+                || fixture.Hard == false
+                || (fixture.CollisionMask & ClimbingCollisionGroup) == 0)
+                continue;
+
+            climbingComp.DisabledFixtures.Add(fixture);
+
+            var climbFixture = new Fixture(fixture.Shape, fixture.CollisionLayer,
+                    fixture.CollisionMask & ~ClimbingCollisionGroup,
+                    true)
+                {ID = $"{ClimbingFixtureName}-{name}"};
+            toAdd.Add(climbFixture);
         }
 
-        private void AddClimbVerb(EntityUid uid, ClimbableComponent component, GetVerbsEvent<AlternativeVerb> args)
+        foreach (var fixture in climbingComp.DisabledFixtures)
         {
-            if (!args.CanAccess || !args.CanInteract || !_actionBlockerSystem.CanMove(args.User))
-                return;
+            _fixtureSystem.DestroyFixture(physicsComp, fixture, manager: fixturesComp);
+        }
+
+        foreach (var fixture in toAdd)
+        {
+            if (!_fixtureSystem.TryCreateFixture(physicsComp, fixture, manager: fixturesComp))
+                return false;
+            climbingComp.Fixtures.Add(fixture.ID);
+        }
 
-            // Check that the user climb.
-            if (!EntityManager.TryGetComponent(args.User, out ClimbingComponent? climbingComponent) ||
-                climbingComponent.IsClimbing)
+        if (!_fixtureSystem.TryCreateFixture(physicsComp,
+            new Fixture(new PhysShapeCircle { Radius = 0.35f }, (int) CollisionGroup.None, ClimbingCollisionGroup, false)
+                {ID = ClimbingFixtureName}, manager: fixturesComp))
+            return false;
+        return true;
+    }
+
+    private void OnClimbEndCollide(EntityUid uid, ClimbingComponent component, EndCollideEvent args)
+    {
+        if (args.OurFixture.ID != ClimbingFixtureName
+            || !component.IsClimbing
+            || component.OwnerIsTransitioning
+            || !TryComp<TransformComponent>(uid, out var transformComp)
+            || !TryComp<PhysicsComponent>(uid, out var physicsComp)
+            || !TryComp<FixturesComponent>(uid, out var fixturesComp))
+            return;
+
+        foreach (var (fixture, _) in args.OurFixture.Contacts)
+        {
+            if (fixture == args.OtherFixture)
+                continue;
```

Comment:
Isn't contacts a dictionary?

Reply:
It is a Dictionary. This loops over all contacts that aren't the current one to check if it should stop climbing since it's not touching any more Climbable entities.